### ISSUE TITLE
[Linux] Native Wayland support

### ISSUE THREAD

**Issue by Enderteck** [neoforged/NeoForge#657]
I wasn't able to find a place to really suggest that in the discord server so, here I go. Hopefully someone sees this.

I know this could be a lot to ask but I'm not here to be technical this is just a wish I have.
Minecraft isn't yet willing to implement this themselves for now:
[Minecraft Feedback Post](https://feedback.minecraft.net/hc/en-us/community/posts/5914275935629-Support-Wayland)

**Would it be possible to detect and use Wayland natively ?**

I think this would probably require GLFW changes but I'm not qualified to know.

This is getting quite important as it's the default for most distros and will be more and more used. 
Xorg is getting more and more outdated and it would also mean some safety risks to continue to only support it.

Native Wayland support is needed and the best solution as far as I know, is modloader wide support. Any other ways create incompatibilities with mods and such.

Any other option such as mods and GLFW patches I've tried didn't work on my Neoforge modpacks (game crashes). Patches and changes in GLFW libs have worked for me on fabric and quilt (with lighter and performance oriented modpacks). 
**Having to change a library not supported by default to use the new default Display protocol natively is annoying.**

Also there are some bugs using Xwayland, such as fullscreen being rendered higher than the screen.
It cuts off the top centimeter of the screen or more. This is fixable but a pain to deal with.

Three screenshots here:

![Screenshot Xwayland 1](https://github.com/neoforged/NeoForge/assets/123917079/62691839-909c-4e77-94d9-d82b808c152a)

![Screenshot Xwayland 2](https://github.com/neoforged/NeoForge/assets/123917079/8d3339aa-fa1f-45a0-b003-b38fff483a6f)

![Screenshot Xwayland 3](https://github.com/neoforged/NeoForge/assets/123917079/447dfe3d-2d29-482e-a9ef-9a7dd14aa55f)

Another problem reported here:

[Issue #393](https://github.com/neoforged/NeoForge/issues/393)

I will also ask Fabric and Quilt to support native wayland at a later date.

The most I've done in technical stuff is downloading artifact builds off of github and changing the GLFW library path. I'm not experienced in Java or Linux Display protocols.

Thanks for any answers or clarifications.

**Comment by marchermans** [comment]
Sadly this is going to be really difficult to implement.
We have now opened up the option for modders to fix this themselves by providing the option to create their own ImmediateWindowProvider, and GraphicsBootstrapper.

However this needs an update of FML in NeoForge

**Comment by Enderteck** [comment]
Would providing a patched GLFW library directly in Forge when detecting Wayland not work ? This would also force miss devs not to break Wayland support as it would be integrated on forge

**Comment by FiniteReality** [comment]
> Would providing a patched GLFW library directly in Forge when detecting Wayland not work ? This would also force miss devs not to break Wayland support as it would be integrated on forge

No, it wouldn't work because the way LWJGL loads its libraries is incredibly cursed. Furthermore, Wayland only supports GLES/EGL which makes it incompatible with most mods already.

**Comment by Enderteck** [comment]
I did a bit of Googling and I have more questions if you don't mind answering them.

Why can I use a patched version of GLFW to play natively on wayland with Fabric/Quilt but I can't on forge ?

Would using something like [Holy GL4ES](https://pojavlauncherteam.github.io/RENDERERS.html#holy-gl4es) like in the mobile Minecraft: Java community port be possible ?
Would it result in bad performance ?

Thx for answering my questions and sorry to be so uninformed.

**Comment by IMS212** [comment]
This is overcomplicating by a lot; and also the thing about EGL is exaggerated.

Once LWJGL gets an update to GLFW 3.4 (released 3 days ago), Wayland will be the default system. Just wait.

**Comment by Enderteck** [comment]
Can I build GLFW myself and replace the current one to get native Wayland support ?

**Comment by klaribot** [comment]
I'm not sure if any of us can wait any longer for [Wayland/XWayland and NVIDIA devs to agree on whether to implement implicit sync or explicit sync](https://forums.developer.nvidia.com/t/wayland-information-for-r545-beta-release/214275), as Minecraft on XWayland for high refresh rate displays is practically unusable when the refresh rate of the application does not exactly match the refresh rate of the display for NVIDIA proprietary drivers >=545, among the myriad of other bugs that've been left unaddressed that plague XWayland on NVIDIA. (https://gitlab.freedesktop.org/xorg/xserver/-/issues/1317)

So far, we've been able to get Vanilla, Fabric, and Quilt running flawlessly under Wayland (*the modloaders themselves, untested with additional mods*), via cherry-picked patches from both https://github.com/Admicos/minecraft-wayland and https://github.com/BoyOrigin/glfw-wayland/ for GLFW 3.4. The cherrypicked patchset is available in [this AUR package](https://aur.archlinux.org/packages/glfw-wayland-minecraft-cursorfix).

However, none of these patches seem to cut it for Forge or NeoForge, which seems to simply crash immediately upon not being able to obtain a window position because... that's just not a thing on Wayland, and [I don't](https://github.com/flatpak/xdg-desktop-portal/issues/980) [think it](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/72) [ever](https://gitlab.gnome.org/GNOME/mutter/-/issues/316) [will](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/264) [be](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/247) any time soon.
`java.lang.IllegalStateException: GLFW error before init: [0x1000C]Wayland: The platform does not provide the window position`

Excited to see neoforged/FancyModLoader#68 to address neoforged/NeoForge#393 finally merged into [neoforged/FancyModLoader:main](https://github.com/neoforged/FancyModLoader/tree/main), though! One step closer to ridding of X11 completely!

**Comment by Enderteck** [comment]
Omg this might happen. Nice ! Finally !

**Comment by MattSturgeon** [comment]
How does this interact with/benefit from/depend on [Project Wakefield](https://wiki.openjdk.org/display/wakefield/OpenJDK+Project+Wakefield+-+Wayland+desktop+support+for+JDK+on+Linux)?

**Comment by FiniteReality** [comment]
> How does this interact with/benefit from/depend on [Project Wakefield](https://wiki.openjdk.org/display/wakefield/OpenJDK+Project+Wakefield+-+Wayland+desktop+support+for+JDK+on+Linux)?

Most likely, not at all - from skimming through the code, Wakefield appears to be aimed towards java.awt which isn't used by Minecraft anyway

**Comment by MattSturgeon** [comment]
> Most likely, not at all [...] Wakefield appears to be aimed towards `java.awt`

Wakefield's stated goal is for OpenJDK to get full Wayland support\*, so that may extend _slightly_ beyond awt eventually... Although I guess that still won't benefit Minecraft much seeing as most of its platform dependencies are things like GLFW, etc.

_\* Beginning with "full" XWayland support, and eventually native Wayland support_

**Comment by Niko-sk2x** [comment]
> Furthermore, Wayland only supports GLES/EGL which makes it incompatible with most mods already.

Contrary to what quick Google search says, egl on desktop allows using opengl API just fine (early implementations probably didn't though). It's only the reference compositor implementation that used gl es.

**Comment by FiniteReality** [comment]
> > Furthermore, Wayland only supports GLES/EGL which makes it incompatible with most mods already.
> 
> Contrary to what quick Google search says, egl on desktop allows using opengl API just fine (early implementations probably didn't though). It's only the reference compositor implementation that used gl es.

Yup, you're right - this was explained to me in Discord :sweat_smile:

**Comment by klaribot** [comment]
> > Most likely, not at all [...] Wakefield appears to be aimed towards `java.awt`
> 
> Wakefield's stated goal is for OpenJDK to get full Wayland support\*, so that may extend _slightly_ beyond awt eventually... Although I guess that still won't benefit Minecraft much seeing as most of its platform dependencies are things like GLFW, etc.
> 
> _\* Beginning with "full" XWayland support, and eventually native Wayland support_

Again, it's better to avoid using XWayland backwards compatibility, at least for NVIDIA users, because of how broken XWayland is when it's actually running on wayland with the latest NVIDIA proprietary drivers and given direct control of rendeing infrastructure via `modeset=1` kernel arguments (and potentially `fbdev=1` too but I think that's mostly experiemental support for only within the scope of the linux framebuffer terminal).

Pardon my condescending snarkiness, but unless you want to ship the patches in a custom X.Org server with Explicit Sync worked in so that NVIDIA drivers can properly sync with the X.Org server running within Wayland, using XWayland at all isn't going to work, especially for higher refresh rate displays.

Hence the effort to get Minecraft running natively on Wayland, even if it's still going to be API translation layers provided by Mesa and EGLStreams for OpenGL and/or Vulkan all the way down.

**Comment by Enderteck** [comment]
The weird thing to me is that Fabric and Quilt support Wayland if a patched GLFW library is given.

How is it supported there and not on NoeForge ? 

Wouldn't it be possible to do the same with NeoForge ?

If it could be implemented like Fabric, could the Wayland GLFW Library be bundled and used when needed ?

The question is getting important as Linux now runs on 4% of PCs. Wayland is the default or will be soon on most DEs.

**Comment by Enderteck** [comment]
Any updates on the issue ?

Wayland is gonna be the default in nearly all desktop environments and apps are now supporting it or are actively working on it, even the steam deck's compositor is gonna get Wayland support soon. 
Mojang seems to not really care about the issue but Fabric has implemented compatibility with GLFW libs patched for Wayland

**Comment by humanlyhuman** [comment]
> Any updates on the issue ?
> 
> Wayland is gonna be the default in nearly all desktop environments and apps are now supporting it or are actively working on it, even the steam deck's compositor is gonna get Wayland support soon. Mojang seems to not really care about the issue but Fabric has implemented compatibility with GLFW libs patched for Wayland

for me it works just fine with the aur package glfw-wayland-minecraft-cursorfix (https://github.com/BoyOrigin/glfw-wayland) on 1.20.4, but on 1.20.1  i had to set earlyWindowControl to false in .minecraft/config/fml.toml

**Comment by Enderteck** [comment]
Thanks for the info, I'll try it

**Comment by Enderteck** [comment]
> > Any updates on the issue ?
> > Wayland is gonna be the default in nearly all desktop environments and apps are now supporting it or are actively working on it, even the steam deck's compositor is gonna get Wayland support soon. Mojang seems to not really care about the issue but Fabric has implemented compatibility with GLFW libs patched for Wayland
> 
> for me it works just fine with the aur package glfw-wayland-minecraft-cursorfix (https://github.com/BoyOrigin/glfw-wayland) on 1.20.4, but on 1.20.1 i had to set earlyWindowControl to false in .minecraft/config/fml.toml

OMG thank you that solves everything. Thanks !

It can't be that hard to add that in fml, just disable that rule when running on Wayland.

Thank You for real you solved all my XWayland issues. <3

It fixed:

- The game window not being aligned with the screen (pictures at the top).

- A crash every time I pressed the Quit Game button (an entire JDK 17 crash).

**Comment by klaribot** [comment]
So I see disable the earlyWindowControl seems to have fixed native Wayland launching, along with the patched glfw. That's good!

Small problem though. If I try it; as soon as I resize the window:
![image](https://github.com/neoforged/NeoForge/assets/25907774/673fe388-e6cf-41f4-a6e3-49b8b421d81a)

And it just sort of... Washes white until nothing is visible anymore.

**Comment by Enderteck** [comment]
With the patched GLFW for Wayland I get this kind a of bug where the mouse cursor is shown even when it's supposed to be hidden:

![Screenshot](https://github.com/neoforged/NeoForge/assets/123917079/0858832c-608b-4d07-9299-b39c5737982a)

Any solutions ?

**Comment by humanlyhuman** [comment]
i found another workaround it doesn't require patched glfw just set these environment variables : __GLX_VENDOR_LIBRARY_NAME=mesa MESA_LOADER_DRIVER_OVERRIDE=zink GALLIUM_DRIVER=zink
it runs minecraft via zink also i think you need mesa installed the best part is this works on every version, every loader i tried even pre 1.13

**Comment by humanlyhuman** [comment]
> So I see disable the earlyWindowControl seems to have fixed native Wayland launching, along with the patched glfw. That's good!
> 
> Small problem though. If I try it; as soon as I resize the window: ![image](https://private-user-images.githubusercontent.com/25907774/316275505-673fe388-e6cf-41f4-a6e3-49b8b421d81a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NDMyNzAsIm5iZiI6MTcxMTY0Mjk3MCwicGF0aCI6Ii8yNTkwNzc3NC8zMTYyNzU1MDUtNjczZmUzODgtZTZjZi00MWY0LWE2ZTMtNDliOGI0MjFkODFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI4VDE2MjI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzZDY0OTY2MmJjZjg0NTZkNGE4MGE2ZjQwZjU1MTcxZjczMjU5N2I5MTE5ZTU4YmNlMDU4YmEzM2UxZTM3ZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hW-a3Lld5FmSl433XrX1i96ncuVP9FYc6D_0-ocoW6o)
> 
> And it just sort of... Washes white until nothing is visible anymore.

are you using nouveau?

**Comment by Username404-59** [comment]
> i found another workaround it doesn't require patched glfw just set these environment variables : __GLX_VENDOR_LIBRARY_NAME=mesa MESA_LOADER_DRIVER_OVERRIDE=zink GALLIUM_DRIVER=zink it runs minecraft via zink also i think you need mesa installed the best part is this works on every version, every loader i tried even pre 1.13

Zink hurts performance since it translates OpenGL to Vulkan, especially with like e.g shaders, so this is not ideal

**Comment by klaribot** [comment]
Hey, so remember way back in March when I said "I'm not sure if any of us can wait any longer for [Wayland/XWayland and NVIDIA devs to agree on whether to implement implicit sync or explicit sync](https://forums.developer.nvidia.com/t/wayland-information-for-r545-beta-release/214275)"?

Well, silly me for being impatient, [explicit sync finally made it into XWayland](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/967) [and](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/90) [related](https://invent.kde.org/plasma/kwin/-/merge_requests/4693) [projects](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3300), so maybe this issue can be revisited with XWayland as the stopgap for now.

Personally, I've moved back to X11, as there are plenty of other regressions that have come out of this that have made Wayland unusable for me for a considerable amount of time now.

UPDATE: 1 week later after posting this and getting the Nvidia 560 beta drivers, Wayland suddenly works perfectly again :)

**Comment by humanlyhuman** [comment]
as of nvidia 555 all versions work perfectly for me, forgot about this thread completely

**Comment by humanlyhuman** [comment]
> > i found another workaround it doesn't require patched glfw just set these environment variables : __GLX_VENDOR_LIBRARY_NAME=mesa MESA_LOADER_DRIVER_OVERRIDE=zink GALLIUM_DRIVER=zink it runs minecraft via zink also i think you need mesa installed the best part is this works on every version, every loader i tried even pre 1.13
> 
> Zink hurts performance since it translates OpenGL to Vulkan, especially with like e.g shaders, so this is not ideal

I know at the time I was just happy to see it work and didn't do much testing

**Comment by virtualspan** [comment]
So using either the glfw-wayland or the glfw-git [(AUR)](https://aur.archlinux.org/packages/glfw-git) package which supports wayland on arch, I just simply tested running Fabric on 1.21.4, and it worked, then same with Forge, didn't work, finally, I then tried the same thing with NeoForge and it worked. To be more specific, I installed the package (glfw-git and tested again with glfw-wayland), then enabled "Use system installation of GLFW" in PrismLauncher, launched the game, it worked, then checked if it was running native and without xwayland, and it was native. I'm not sure why NeoForge works and Forge doesn't, but I guess that's another reason NeoForge is better.

Edit: I was using the latest NeoForge beta (21.4.86-beta at the time of writing)
Edit 2: I tested ATM10 (which uses NeoForge) and it worked out-of-the-box. I've had no issues or differences compared to xwayland on all the tests I've mentioned above (obviously excluding regular Forge).
Edit 3: Even older version of NeoForge like version 21.1.1 on 1.21.1 work. But versions too old like 47.1.106 on 1.20.1 don't work.

**Comment by virtualspan** [comment]
Seems like older versions of NeoForge work too, I tested BMC4 with Neoforge version 47.1.106 on 1.20.1 and it worked

**Comment by Renari** [comment]
Interesting I've tried a setup similar to @virtualspan with very different results: https://mclo.gs/dtYej13
```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: There is no OpenGL context current in the current thread.
	at MC-BOOTSTRAP/fml_earlydisplay@4.0.42/net.neoforged.fml.earlydisplay.DisplayWindow.setupMinecraftWindow(DisplayWindow.java:535) ~[earlydisplay-4.0.42.jar%23148!/:4.0] {re:mixin}
	at MC-BOOTSTRAP/fml_loader@4.0.42/net.neoforged.fml.loading.ImmediateWindowHandler.setupMinecraftWindow(ImmediateWindowHandler.java:77) ~[loader-4.0.42.jar%23149!/:4.0] {}
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.platform.Window.mixinextras$bridge$setupMinecraftWindow$71(Window.java) ~[client-1.21.1-20240808.144430-srg.jar%23566!/:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:mixin:APP:sodium-common.mixins.json:core.WindowMixin from mod sodium,pl:mixin:APP:sodium-common.mixins.json:workarounds.context_creation.WindowMixin from mod sodium,pl:mixin:APP:immediatelyfast-common.mixins.json:core.MixinWindow from mod immediatelyfast,pl:mixin:APP:relics.mixins.json:WindowMixin from mod relics,pl:mixin:APP:kubejs.mixins.json:WindowMixin from mod kubejs,pl:mixin:APP:mixins.iris.json:MixinWindow from mod iris,pl:mixin:A,pl:runtimedistcleaner:A}
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.platform.Window.wrapOperation$cbd000$sodium$wrapGlfwCreateWindowForge(Window.java:1087) ~[client-1.21.1-20240808.144430-srg.jar%23566!/:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:mixin:APP:sodium-common.mixins.json:core.WindowMixin from mod sodium,pl:mixin:APP:sodium-common.mixins.json:workarounds.context_creation.WindowMixin from mod sodium,pl:mixin:APP:immediatelyfast-common.mixins.json:core.MixinWindow from mod immediatelyfast,pl:mixin:APP:relics.mixins.json:WindowMixin from mod relics,pl:mixin:APP:kubejs.mixins.json:WindowMixin from mod kubejs,pl:mixin:APP:mixins.iris.json:MixinWindow from mod iris,pl:mixin:A,pl:runtimedistcleaner:A}
```

`glfw-git` has the same issue

What I've done to try and force wayland:
```
__EGL_VENDOR_LIBRARY_FILENAMES=/usr/share/glvnd/egl_vendor.d/10_nvidia.json
__GLX_VENDOR_LIBRARY_NAME=nvidia
__NV_PRIME_RENDER_OFFLOAD=1
__VK_LAYER_NV_optimus=NVIDIA_only
GBM_BACKEND=nvidia-drm
LIBVA_DRIVER_NAME=nvidia
GLFW_PLATFORM=wayland
LWJGL_PLATFORM=wayland
```
As well as the aforementioned use native library checkboxes for GLFW. None of these seem to make a difference.

**Comment by sciwhiz12** [comment]
In [Minecraft 26.3-snapshot-4](https://www.minecraft.net/en-us/article/minecraft-26-3-snapshot-4), Mojang switched from GLFW to SDL3 for "window management, input and platform integration." Due to this, Minecraft itself will support running on Wayland natively once 26.3 lands.

I don't think it's worth the trouble for us to implement and maintain Wayland support, now that Minecraft itself has done the move for us. It's only a matter of waiting until 26.3 lands sometime later this year.

(If anyone can find and make a PR implementing a good solution that won't be a maintenance pain or make us change libraries, I of course welcome your PR.)

On that basis, I'll be closing this issue as won't fix.
